cilium/CFP-34841-bgp-route-learning.md

```diff
@@ -0,0 +1,118 @@
+
+#### Encapsulation Mode
+
+The author requests guidance from the community to support this feature in Encapsulation mode.
+
+
+### BGP Import Policy
+Under [this PR](https://github.com/cilium/cilium/pull/33035), I modified Cilium to reject all BGP paths advertised toward it.  This proposal will revert the majority of those changes.
+
+
+### Custom Resource Definition Modifications
+
+To mitigate the risk associated with route leaks, a new configuration option named `prefixLimit` will be added to `CiliumBGPPeeringPolicy` and `CiliumBGPPeerConfig`.  Exceeding the configured prefix limit will result in the the BGP speaker being torn down.  Under the covers, GoBGP supports additional configuration for this option as seen in [link](https://pkg.go.dev/github.com/osrg/gobgp/internal/pkg/config#PrefixLimitConfig).  It may be desirable to expose an equivalent CRD configuration option for GoBGP's `RestartTimer`.
+
+
+## Impacts / Key Questions
+
+_List crucial impacts and key questions. They likely require discussion and are required to understand the trade-offs of the CFP. During the lifecycle of a CFP, discussion on design aspects can be moved into this section. After reading through this section, it should be possible to understand any potentially negative or controversial impact of this CFP. It should also be possible to derive the key design questions: X vs Y._
+
+### Impact: Cilium complexity
+
+Implementation of this feature introduces a fair amount of complexity to Cilium.  Complex systems are harder to reason about.  Complexity introduced includes a second BGP daemon, manipulation of the node's Linux routing table, and additional configuration options.
+
+Further, to support this feature in Encapsulation routing mode may require extensive changes outside of BGP-related source code.
+
+### Key Question: Should the feature be limited to Native-Routing mode?
+
+When in Native-Routing mode, the scope of changes proposed are limited to the BGP-related source code and the introduction of a second BGP daemon.
+
+### Key Question: Does the introduction of the feature REQUIRE the administrator to advertise a default route via BGP?
+
+With the existing behavior relying on a default route, one not learned via BGP, does the introduction of the feature remove the ability to use a default route?  Does it require the adminstrator to announce a default route via BGP instead?
+
+### Option 1:  Implement the Proposed Feature
+
+#### Pros
+
+* Unlocks new use-cases for Cilium with advanced network designs.
+
+#### Cons
+
+* Introduces additional complexity for both maintainers and administrators.  Additional complexity for administrators resides in the fact that the underlying Linux routing table may change without notice in response to BGP routing changes.
+
+### Option 2: Do Nothing
```

Comment:
I think we do need to enhance Cilium's capability to install learned BGP routes. I'd add 3rd option in this CFP as well, that is to pass responsibility of installing kernel routes to Cilium itself instead of another process. 

There are few advantages to it
Pros : 
- Lifecycle and management of BGP on the node is via Cilium.  ( Although in some scenarios this might be a drawback as well )
- Tighter control over what gets installed in kernel and with which priority ( admin-distance ) when there are multiple sources of the route ( BGP instances, or other Cilium features such as auto-direct-node-routing )

Cons
- Complex to implement ( Essentially we have to implement RIB engine inside Cilium ).
- Cilium would need to sync kernel routes on restarts.

Reply:
I had assumed that implementing RIB management directly into Cilium would deviate too far from its intended scope.  If we were to invest the time to implement RIB programming, I'd want to ask "Why not invest that time into implementing it in GoBGP directly" instead?  cc @YutaroHayakawa

Aside from the pros/cons listed, would native RIB programming within Cilium unblock new features, or reduce existing tech debt somehow?

Reply:
> "Why not invest that time into implementing it in GoBGP directly"

For me the answer for this question is "Eventually, we may want to go this direction, but it's too ambitious for now". I think there is a huge gap between having the generic RIB implementation for GoBGP and implementing a RIB specific to the Cilium's use case.

To decouple the RIB implementation from Cilium, we need to come up with a stable interface between the RIB and Cilium. Designing such an interface is already hard. Putting them into a single Go binary would simplifies the implementation a lot. We can break the interface at any time as needed.

Once our implementation become mature enough, we can always consider extracting our implementation to the independent project. However, we shouldn't set it as a goal from the beginning.

> Aside from the pros/cons listed, would native RIB programming within Cilium unblock new features, or reduce existing tech debt somehow?

It allows us to program the Cilium's eBPF data plane. That leaves us the possibility to implement the data plane feature that doesn't exist in the Linux kernel with eBPF and integrate it with BGP. If we go with the GoBGP-oriented approach, it is hard to justify the support for such a specialized data plane.

===

As a one engineer, I agree with you. GoBGP should get a proper generic RIB and data plane manager implementation (like Zebra), but I guess it's hard to do that in the context of the current Cilium project.

Reply:
<img width="60" alt="Screenshot 2024-10-03 at 7 08 13 PM" src="https://github.com/user-attachments/assets/c0aed16f-62ac-4443-8eef-14ab5e59c172">  @YutaroHayakawa!